packages/ember-data/lib/serializers/rest-serializer.js

```diff
@@ -689,6 +689,49 @@ var RESTSerializer = JSONSerializer.extend({
     } else {
       json[key + "Type"] = Ember.String.camelize(belongsTo.modelName);
     }
+  },
+
+  /**
+    You can use this method to customize how a polymorphic relationship should
+    be extracted.
+
+    @method extractPolymorphicRelationship
+    @param {Object} relationshipModelName
+    @param {Object} relationshipHash
+    @param {Object} relationshipOptions
+    @return {Object}
+   */
+  extractPolymorphicRelationship: function(relationshipModelName, relationshipHash, relationshipOptions) {
+    var { relationshipKey, resourceHash, relationshipMeta } = relationshipOptions;
+
+    // A polymorphic belongsTo relationship can be present in the payload
+    // either in the form where the `id` and the `type` are given:
+    //
+    //   {
+    //     message: { id: 1, type: 'post' }
+    //   }
+    //
+    // or by the `id` and a `<relationship>Type` attribute:
+    //
+    //   {
+    //     message: 1,
+    //     messageType: 'post'
+    //   }
+    //
+    // The next code checks if the latter case is present and returns the
+    // corresponding JSON-API representation. The former case is handled within
+    // the super class (JSONSerializer).
+    var isPolymorphic = relationshipMeta.options.polymorphic;
+    var typeProperty = `${relationshipKey}Type`;
```

Comment:
Just to make sure, my intention here was to do ~~`keyForAttribute`~~ `keyForRelationship` on the `typeProperty` string.

Even though `relationshipKey` is cased correctly that doesn't mean we can just append `Type` to the end of it.

Say you have a `bestFriend: belongsTo()` and you've overriden `keyForRelationship` to returned dasherized. `bestFriend` is now `best-friend`. Odds are your payload contains `best-friend-type`, not `best-friendType`.